Add a simple accelerator selection mechanism.

zolkis opened the pull request:
Fixes #815 

As explained in #884, add context options and attributes/internal slots that can be used for conveying application hints wrt the preferred acceleration type (CPU or massively parallel processing, i.e. NPU or GPU).

This is a minimal change, and we might want to refine more the algorithms wrt. context power preferences and acceleration options (currently not addressed). These could be done in this PR, or in a separate subsequent PR.

handellm commented:
Seems good!

huningxin commented on index.bs:
```diff
@@ -1178,7 +1216,8 @@ Note: `dispatch()` itself provides no signal that graph execution has completed.
     1. Enqueue the following steps to |graph|.{{MLGraph/[[context]]}}.{{MLContext/[[timeline]]}}:
         1. Run these steps, but [=/abort when=] [=this=] [=MLContext/is lost=]:
-            1. Issue a compute request to |graph|.{{MLGraph/[[implementation]]}} given |inputs| and |outputs|.
+            1. Issue a compute request to |graph|.{{MLGraph/[[implementation]]}} given |inputs| and |outputs|, as well as |graph|.{{MLGraph/[[context]]}}.{{MLContext/[[powerPreference]]}} and |graph|.{{MLGraph/[[context]]}}.{{MLContext/[[accelerated]]}}.
```
I suppose `powerPreference` and `accelerated` options should be used by `build` steps rather than `dispatch`?

huningxin commented on index.bs:
```diff
@@ -1095,6 +1105,12 @@ interface MLContext {
+        The {{MLContext}}'s processing type (CPU or massively parallel processing).
+    : <dfn>\[[cpuFallbackActive]]</dfn> of type {{boolean}}.
+    ::
+        The {{MLContext}}'s status for CPU fallback type (CPU or massively parallel processing).
```
AFAIK, the major native ML runtimes, including Core ML, Windows ML (ONNX Runtime) and TFLite, commonly enable CPU fallback by default. Some runtimes, e.g. ONNX Runtime, allow developers to disable CPU fallback explicitly through a session option `disable_cpu_ep_fallback` [1]. What is the case that a WebNN implementation may set this to false?

[1]: https://onnxruntime.ai/docs/genai/reference/config.html#sessionoptions

zolkis replied:
That was discussed in earlier calls (in the explainer related discussions): exposing a context option for setting CPU fallback to false hits some constraints and could be accomplished with the `accelerated` option, hence was discarded as an approach.

In #884 there is a code example for this use case:

```js
// create a context that should use massive parallel processing (e.g. GPU/NPU)
context = await navigator.ml.createContext({accelerated: true});
if (context.accelerated) {
    // the context will mostly use GPU/NPU, but CPU fallback may happen
} else {
    // the platform tells it likely cannot provide NPU or GPU, so try something else
}

// create a context that should preferably use NPU
context = await navigator.ml.createContext({accelerated: true, powerPreference: 'low-power'});
if (context.accelerated) {
    // NPU is likely used -- further requirements could be set by opSupportLimitsPerDevice
} else {
    // NPU is likely not available, and since GPU needs high power, it is not used
}
```

anssiko replied:
This sounds reasonable to me.

@zolkis if you agree and are available, please open a separate issue for `cpuFallbackActive` and seed it with your insights. If you also update this PR accordingly we should be able to merge this PR by the end of the week.

zolkis replied:
> `cpuFallbackActive` seems to be a useful attribute of `MLGraph`

I agree, that makes a lot of sense. Even more, actually a sub-graph or individual ops might fall back to CPU (as mentioned before, a context / graph should be associated with an execution plan, not only with underlying execution devices).

For this PR, exposing `cpuFallbackActive` on context was chosen for the "simplicity" argument, also because a context still is associated with an underlying execution device -- for which we opened another issue in #897. Once we relax that and work with [these terms](https://github.com/webmachinelearning/webnn/issues/897#issuecomment-3455923178), I think we should properly address CPU fallback as well.

This is a good development, so I will remove it from this PR.

zolkis commented on index.bs:
```diff
@@ -1178,7 +1216,8 @@ Note: `dispatch()` itself provides no signal that graph execution has completed.
         1. Run these steps, but [=/abort when=] [=this=] [=MLContext/is lost=]:
-            1. Issue a compute request to |graph|.{{MLGraph/[[implementation]]}} given |inputs| and |outputs|.
+            1. Issue a compute request to |graph|.{{MLGraph/[[implementation]]}} given |inputs| and |outputs|, as well as |graph|.{{MLGraph/[[context]]}}.{{MLContext/[[powerPreference]]}} and |graph|.{{MLGraph/[[context]]}}.{{MLContext/[[accelerated]]}}.
+            1. Run the steps to [=poll CPU fallback status=] for |graph|.{{MLGraph/[[context]]}}.
```
Right, these would only be needed if there was an event (discussed earlier and agreed that polling is enough for now).

handellm commented on index.bs:
```diff
@@ -1114,6 +1130,30 @@ The <dfn>context type</dfn> is the type of the execution context that manages th
+  </summary>
+    1. If [=this=].{{MLContext/[[accelerated]]}} is `false`, then:
+        1. Set [=this=].{{MLContext/[[cpuFallbackActive]]}} to `true` and return.
+    1. Issue a request to check whether the underlying platform uses CPU as the main underlying execution device for inference. If yes, then implementations <span class=allow-2119>should</span> set [=this=].{{MLContext/[[cpuFallbackActive]]}} to `true` and return.
```
Is "issue a request" achievable in a synchronous manner? If not perhaps this should be an async method.

zolkis commented:
@huningxin your feedbacks have been addressed:
- removed `cpuFallbackActive` from this PR (for later introduction);
- moved the `accelerated` consideration from the `dispatch()` steps to the `build()` steps.
When merging this PR, please keep the `cpuFallbackActive` commit in history.

huningxin reviewed (approved):
LGTM, thanks @zolkis !

anssiko commented:
I will now merge this and open a new issue for more narrowly scoped discussion wrt the CPU fallback mechanism.

Again, thank you @zolkis and everyone who contributed to this feature through issue discussions, design proposals and review. This establishes a baseline for future device selection hints and enhancements.

anssiko commented:
@fdwr, thank you, this is great input. I agree this is analogous to font selection.

Would you mind opening a new issue for this, seed it with your comment? I'd like to put this on our F2F agenda to capture feedback on use case-driven priorities.
